### PR TITLE
Use coreToLg in Generate

### PR DESCRIPTION
**Pull request by nikivazou**

### CODE REVIEW

File: src/Language/Haskell/Liquid/Constraint/Generate.hs [hunk 1]
```diff
@@ -1333,26 +1333,11 @@ argExpr γ (App e (Type _)) = argExpr γ e
...
-  where allowTC = typeclass (getConfig γ)
-lamExpr γ (Let (NonRec x ex) e) = case (lamExpr γ ex, lamExpr γ e) of
-                                       (Just px, Just p) -> Just (p `F.subst1` (F.symbol x, px))
-                                       _  -> Nothing
-lamExpr γ (Lam x e)   = case lamExpr γ e of
-                            Just p -> Just $ F.ELam (F.symbol x, typeSort (emb γ) $ Ghc.expandTypeSynonyms $ varType x) p
-                            _ -> Nothing
-lamExpr _ _           = Nothing
-
+lamExpr g e = case runToLogic (emb g) mempty mempty (\x -> todo Nothing ("coreToLogic not working lamExpr" ++ x)) (coreToLogic True e) of
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Constraint/Generate.hs]:
```suggestion
lamExpr g e = case runToLogic (emb g) mempty mempty (\x -> todo Nothing ("coreToLogic not working lamExpr: " ++ x)) (coreToLogic True e) of 
```

File: tests/todo/T2076.hs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module T2076 where
```
Comment (facundominguez) [inline tests/todo/T2076.hs:1]: Should this be placed outside of `tests/todo` if it passes?

Also, this needs to be listed in `tests/tests.cabal` for CI to pick it up.

Comment (nikivazou) [reply tests/todo/T2076.hs:1]: It is not passing here, but it is passing in your branch (with your edits).

**Comment by facundominguez** [comment]
I explored [the failure](https://app.circleci.com/pipelines/github/ucsd-progsys/liquidhaskell/1523/workflows/44ffe0a4-c48a-40df-b768-cac64b930994/jobs/8379?invite=true#step-112-3146) in CI today.
```
[20 of 24] Compiling NaturalDeduction ( benchmarks/popl18/nople/pos/NaturalDeduction.hs, /home/circleci/project/dist-newstyle/build/x86_64-linux/ghc-8.10.7/tests-0.1.0.0/x/prover-nople-pos/build/prover-nople-pos/prover-nople-pos-tmp/NaturalDeduction.o )

<no location info>: error:
    :1:1-1:1: Error
  Constraint with free vars
  50
  [lqdc##$select##NaturalDeduction.PAnd##1,
   lqdc##$select##NaturalDeduction.PAnd##2]
  Try using the --prune-unsorted flag
```

The selectors are absent from the `.fq` file both before and after the changes in this PR. 
Adding `--exact-data-cons` has the selectors added with slightly different names, so they still aren't found. We get:
```
constant NaturalDeduction.PAnd##lqdc##$select##NaturalDeduction.PAnd##1 : (func(2 , [(NaturalDeduction.PAnd @(0) @(1));
                                                                                     @(0)]))
constant NaturalDeduction.PAnd##lqdc##$select##NaturalDeduction.PAnd##2 : (func(2 , [(NaturalDeduction.PAnd @(0) @(1));
                                                                                     @(1)]))
```

@nikivazou, would this bring some insight on how to fix the test?

File: src/Language/Haskell/Liquid/Constraint/Generate.hs [hunk 3]
```diff
@@ -1333,26 +1333,11 @@ argExpr γ (App e (Type _)) = argExpr γ e
...
-  where allowTC = typeclass (getConfig γ)
-lamExpr γ (Let (NonRec x ex) e) = case (lamExpr γ ex, lamExpr γ e) of
-                                       (Just px, Just p) -> Just (p `F.subst1` (F.symbol x, px))
-                                       _  -> Nothing
-lamExpr γ (Lam x e)   = case lamExpr γ e of
-                            Just p -> Just $ F.ELam (F.symbol x, typeSort (emb γ) $ Ghc.expandTypeSynonyms $ varType x) p
-                            _ -> Nothing
-lamExpr _ _           = Nothing
-
+lamExpr g e = case runToLogic (emb g) mempty mempty (\x -> todo Nothing ("coreToLogic not working lamExpr: " ++ x)) (coreToLogic True e) of
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Constraint/Generate.hs]: Further inspection of the code reveals that `runToLogic` would only produce selectors that start with `lqdc`, unless they are provided in the third argument of `runToLogic` with a different symbol.

**Comment by facundominguez** [comment]
I'm trying a fix that gets `NaturalDeduction.hs` passing verification, but I have the impression that the second argument of `runToLogic` of type `LogicMap` also needs to be set. Looks like it would need to be passed all the way down from `Plugin.hs`.

**Comment by nikivazou** [comment]
Nice! I saw it, but could not reproduce the crash locally, so got stuck with that. 
Maybe it is because I am using the executable instead of the plugin. 
I will go back into it by Monday (unless you already fix it :))

**Comment by facundominguez** [comment]
There is another failure in CI:
```
[7 of 9] Compiling STMonad          ( classes/pos/STMonad.hs, /home/circleci/project/dist-newstyle/build/x86_64-linux/ghc-8.10.7/tests-0.1.0.0/x/class-pos/build/class-pos/class-pos-tmp/STMonad.o, /home/circleci/project/dist-newstyle/build/x86_64-linux/ghc-8.10.7/tests-0.1.0.0/x/class-pos/build/class-pos/class-pos-tmp/STMonad.dyn_o )

**** LIQUID: ERROR *************************************************************

<no location info>: error:
    :1:1-1:1: Error
  elaborate solver elabBE 233 "lq_anf$##7205759403792808725##d3VX" {VV : func(0 , [int; (STMonad.ST int int)]) | [(VV =
                                                  lam lam_arg##1 : int . (GHC.Base.return STMonad.$fMonadST (STMonad.checkEq 3 (lam_arg##1  :  int))))]} failed on:
      VV == lam lam_arg##1 : int . GHC.Base.return STMonad.$fMonadST (STMonad.checkEq 3 (lam_arg##1 : int))
  with error
      The sort (@(3167) (GHC.Base.Monad (STMonad.ST @(3169)))) is not a function with at least 1 arguments

  in environment
      GHC.Base.return := func(2 , [@(1); (@(0) @(1))])

      STMonad.$fMonadST := func(1 , [(GHC.Base.Monad (STMonad.ST @(0)))])

      STMonad.checkEq := func(0 , [int; int; int])

      VV := func(0 , [int; (STMonad.ST int int)])
```
This appears with the first commit in this PR. I've checked the `.fq` files before and after the change. In essence, after the change, the dictionary of `Monad (ST s)` is present in the terms.

Before the change we get a lambda expression like:
```
lq_anf$ : {VV : func(0 , [int; (STMonad.ST int int)]) | [(VV =
  lam n0##a1Ew : int .
    (GHC.Base.$62$$62$$61$
       STMonad.fresh
       lam n1##a1Ex : int .
         (GHC.Base.$62$$62$$61$
           STMonad.fresh
           lam n2##a1Ey : int .
             (GHC.Base.$62$$62$$61$
               STMonad.fresh
               lam n3##a1Ez : int .
                 (GHC.Base.return
                   (STMonad.checkEq 3 n3##a1Ez))))))]}
```
And after the change we get
```
lq_anf$ : {VV : func(0 , [int; (STMonad.ST int int)]) | [(VV =
  lam n0##a1Ew : int .
    (GHC.Base.$62$$62$$61$
       STMonad.$fMonadST
       STMonad.fresh
       lam n1##a1Ex : int .
         (GHC.Base.$62$$62$$61$
           STMonad.$fMonadST
           STMonad.fresh
           lam n2##a1Ey : int .
             (GHC.Base.$62$$62$$61$
               STMonad.$fMonadST
               STMonad.fresh
               lam n3##a1Ez : int .
                 (GHC.Base.return
                   STMonad.$fMonadST
                   (STMonad.checkEq 3 n3##a1Ez))))))]}
```

The test starts passing if I flip the invocation of `coreToLogic True e` to `coreToLogic False e`. But I don't grasp the consequences of flipping `allowTC` in the invocation of `coreToLogic`.

**Comment by facundominguez** [comment]
> But I don't grasp the consequences of flipping `allowTC` in the invocation of `coreToLogic`.

In 478751d I'm setting `allowTC` only if it is set in the configuration.

**Pull request merged**